xfrd-catalog-zones.c

```diff
@@ -0,0 +1,1275 @@
+
+	else if (!xfrd->catalog_consumer_zones)
+		msg = "zone not found: "
+		      "xfrd's catalog consumer zones index is empty";
+
+#ifndef MULTIPLE_CATALOG_CONSUMER_ZONES
+	else if (xfrd->catalog_consumer_zones->count > 1)
+		return "not processing: more than one catalog consumer zone "
+		       "configured and only a single one allowed";
+#endif
+	else if (!(consumer_zone = (struct xfrd_catalog_consumer_zone*)
+	         rbtree_search(xfrd->catalog_consumer_zones, zone->node.key)))
+		msg = "zone not found in xfrd's catalog consumer zones index";
+	else
+		return consumer_zone->invalid;
+
+	if (msg)
+		log_msg(LOG_ERR, "catalog consumer zone '%s': %s",
+				zone->name, msg);
+
+	return msg;
+}
+
+void xfrd_process_catalog_consumer_zones()
+{
+#ifndef MULTIPLE_CATALOG_CONSUMER_ZONES
+	xfrd_process_catalog_consumer_zone(xfrd_one_catalog_consumer_zone());
+#else
+	struct xfrd_catalog_consumer_zone* consumer_zone;
+
+	RBTREE_FOR(consumer_zone, struct xfrd_catalog_consumer_zone*,
+			xfrd->catalog_consumer_zones) {
+		xfrd_process_catalog_consumer_zone(consumer_zone);
+	}
+#endif
+}
+
+static inline struct catalog_member_zone* cursor_cmz(rbnode_type* node)
+{ return node != RBTREE_NULL ? (struct catalog_member_zone*)node->key : NULL; }
+static inline const dname_type* cursor_member_id(rbnode_type* node)
+{ return cursor_cmz(node) ? cursor_cmz(node)->member_id : NULL; }
+
+#if !defined(NDEBUG) && 1 /* Only disable for seriously slow debugging */
+static void debug_log_consumer_members(
+		struct xfrd_catalog_consumer_zone* consumer_zone)
+{
+	rbnode_type* cursor;
+	size_t i;
+
+	for ( cursor = rbtree_first(&consumer_zone->member_ids), i = 0
+	    ; cursor != RBTREE_NULL; i++, cursor = rbtree_next(cursor)) {
+		DEBUG(DEBUG_XFRD,1, (LOG_INFO, "Catalog member %.2zu: %s = %s",
+		      i, dname_to_string(cursor_member_id(cursor), NULL),
+		      cursor_cmz(cursor)->options.name));
+	}
+}
+#else
+# define debug_log_consumer_members(x) /* nothing */
+#endif
+
+static void
+xfrd_process_catalog_consumer_zone(
+		struct xfrd_catalog_consumer_zone* consumer_zone)
+{
+	zone_type* zone;
+	const dname_type* dname;
+	domain_type *match, *closest_encloser, *member_id, *group;
+	rrset_type *rrset;
+	size_t i;
+	uint8_t version_2_found;
+	/* Currect catalog member zone */
+	rbnode_type* cursor;
+	struct pattern_options *default_pattern = NULL;
+	/* A transfer of a catalog zone can contain deletion and adding of
+	 * the same member zone. In such cases it can occur that the member
+	 * is tried to be added before it is deleted. For these exceptional
+	 * cases, we will rewalk the zone after the first pass, to retry
+	 * adding those zones.
+	 *
+	 * Initial pass is mode "try_to_add".
+	 * If a zone cannot be added, mode is set to "retry_to_add"
+	 * If after the first pass the mode is "retry_to_add",
+	 *    mode will be set to "just_add", and a second pass is done.
+	 */
+	enum { try_to_add, retry_to_add, just_add } mode;
+
+	if (!consumer_zone)
```

Comment:
Shouldn't this be an `assert`?

Reply:
There is:
`xfrd_process_catalog_consumer_zone(xfrd_one_catalog_consumer_zone());` on line 439 above. `xfrd_one_catalog_consumer_zone()` can return NULL. I can make and if there and turn this if your want. Do you want that?

Reply:
Addressed in ac967745